src/mesh_smoothing.rs

```diff
@@ -0,0 +1,582 @@
+            faces.clone(),
+            vertices.clone(),
+            NormalStrategy::Sharp,
+        );
+        let relaxed_geometry_1 = laplacian_smoothing(&geometry, 1);
+        let relaxed_geometry_10 = laplacian_smoothing(&geometry, 10);
+
+        assert_eq!(
+            relaxed_geometry_1.triangle_faces_len(),
+            geometry.triangle_faces_len()
+        );
+        assert_eq!(
+            relaxed_geometry_10.triangle_faces_len(),
+            geometry.triangle_faces_len()
+        );
+        assert_eq!(
+            relaxed_geometry_1.vertices().len(),
+            geometry.vertices().len()
+        );
+        assert_eq!(
+            relaxed_geometry_10.vertices().len(),
+            geometry.vertices().len()
+        );
+        assert_eq!(relaxed_geometry_1.normals().len(), geometry.normals().len());
+        assert_eq!(
+            relaxed_geometry_10.normals().len(),
+            geometry.normals().len()
+        );
+    }
+
+    #[test]
+    fn test_laplacian_smoothing() {
+        let (faces, vertices) = triple_torus();
+        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
+            faces.clone(),
+            vertices.clone(),
+            NormalStrategy::Sharp,
+        );
+
+        let (faces_1_i, vertices_1_i) = triple_torus_laplacian_1_iteration();
+        let test_geometry_1_i = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
+            faces_1_i.clone(),
+            vertices_1_i.clone(),
+            NormalStrategy::Sharp,
+        );
+
+        let (faces_3_i, vertices_3_i) = triple_torus_laplacian_3_iterations();
+        let test_geometry_3_i = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
+            faces_3_i.clone(),
+            vertices_3_i.clone(),
+            NormalStrategy::Sharp,
+        );
+
+        let relaxed_geometry_1_i = laplacian_smoothing(&geometry, 1);
+        let relaxed_geometry_3_i = laplacian_smoothing(&geometry, 3);
+
+        let relaxed_geometry_1_i_faces: Vec<TriangleFace> =
+            relaxed_geometry_1_i.triangle_faces_iter().collect();
+        let test_geometry_1_i_faces: Vec<TriangleFace> =
+            test_geometry_1_i.triangle_faces_iter().collect();
+        let relaxed_geometry_3_i_faces: Vec<TriangleFace> =
+            relaxed_geometry_3_i.triangle_faces_iter().collect();
+        let test_geometry_3_i_faces: Vec<TriangleFace> =
+            test_geometry_3_i.triangle_faces_iter().collect();
+
+        assert_eq!(relaxed_geometry_1_i_faces, test_geometry_1_i_faces);
+        assert_eq!(relaxed_geometry_3_i_faces, test_geometry_3_i_faces);
+
+        const TOLERANCE_SQUARED: f32 = 0.01 * 0.01;
+
+        let relaxed_geometry_1_i_vertices = relaxed_geometry_1_i.vertices();
+        let test_geometry_1_i_vertices = test_geometry_1_i.vertices();
+
+        for i in 0..test_geometry_1_i_vertices.len() {
+            assert!(
```

Comment:
I think the `distance_squared` asserts here should become `approx` asserts: https://docs.rs/approx/0.3.2/approx/
We already have `approx` as a dependency.

From the top of my head, I don't know whether to test for relative difference or ULPs, but we should probably pick one of these strategies and stick with it.

Same for every `distance_squared` assert from here to the end of file.